pkg/apis/serving/v1beta1/inference_service_status.go

```
@@ -383,3 +388,102 @@ func (ss *InferenceServiceStatus) ClearCondition(conditionType apis.ConditionTyp
 		conditionSet.Manage(ss).ClearCondition(conditionType)
 	}
 }
+
+func (ss *InferenceServiceStatus) UpdateModelRevisionStates(modelState ModelState, info *FailureInfo) {
+	if ss.ModelStatus.ModelRevisionStates == nil {
+		ss.ModelStatus.ModelRevisionStates = &ModelRevisionStates{ActiveModelState: modelState}
+	} else {
+		ss.ModelStatus.ModelRevisionStates.ActiveModelState = modelState
+	}
+	// Update transition status, failure info based on new model state
+	if modelState == Pending || modelState == Loading {
+		ss.ModelStatus.TransitionStatus = InProgress
+	} else if modelState == Loaded {
+		ss.ModelStatus.TransitionStatus = UpToDate
+	} else if modelState == FailedToLoad {
+		ss.ModelStatus.TransitionStatus = BlockedByFailedLoad
+	}
+	if info != nil {
+		ss.SetModelFailureInfo(info)
+	}
+}
+
+func (ss *InferenceServiceStatus) UpdateModelTransitionStatus(status TransitionStatus, info *FailureInfo) {
+	ss.ModelStatus.TransitionStatus = status
+	// Update model state to 'FailedToLoad' in case of invalid spec provided
+	if ss.ModelStatus.TransitionStatus == InvalidSpec {
+		ss.UpdateModelRevisionStates(FailedToLoad, info)
```

This will overwrite `TransitionStatus` to be `BlockedByFailedLoad` which I don't think is what we want here?

I'm not sure it makes sense to update the model state here at all actually since an invalid spec doesn't affect the active model state, just means we can't fully reconcile.

---

It does not affect the active model, but invalid spec would result in model deployment failure and we should surface this error back to the user right?

---

Yes it will still be reflected in the transitionStatus of `InvalidSpec` and the `failureInfo`. I thought the distinction made sense since it isn't a model loading failure per se, the load isn't even attempted because the spec is invalid.

---

@Suresh-Nakkeran I agree here that we can remove the code updating to `FailedToLoad`, the model loading has not started yet.

---

okay @yuzisun